File: libraries/flicker/src/com/android/server/wm/flicker/monitor/TransitionMonitor.java
/*
 * Copyright (C) 2018 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.server.wm.flicker.monitor;

import android.os.Environment;

import com.google.common.io.BaseEncoding;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

/** Collects test artifacts during a UI transition. */
public abstract class TransitionMonitor {
    static final String TAG = "FLICKER";
    public static final Path OUTPUT_DIR =
            Paths.get(Environment.getExternalStorageDirectory().toString(), "flicker");
    protected String mChecksum;
    protected Path mOutputPath;

    /** Starts monitor. */
    public abstract void start();

    /** Stops monitor. */
    public abstract void stop();

    /**
     * Saves any monitor artifacts to file adding {@code testTag} and {@code iteration} to the file
     * name.
     *
     * @param testTag suffix added to artifact name
     * @param iteration suffix added to artifact name
     * @return Path to saved artifact
     */
    public Path save(String testTag, int iteration) {
        return save(testTag + "_" + iteration);
    }

    /**
     * Saves any monitor artifacts to file adding {@code testTag} to the file name.
     *
     * @param testTag suffix added to artifact name
     * @return Path to saved artifact
     */
    /**
     * Saves trace file to the external storage directory suffixing the name with the testtag and
     * iteration.
     *
     * <p>Moves the trace file from the default location via a shell command since the test app does
     * not have security privileges to access /data/misc/wmtrace.
     *
     * @param testTag suffix added to trace name used to identify trace
     * @return Path to saved trace file and file checksum (SHA-256)
     */
    public Path save(String testTag) {
        mOutputPath.toFile().mkdirs();
        Path savedTrace = saveTrace(testTag);
        mChecksum = calculateChecksum(savedTrace);
        return savedTrace;
    }

    protected Path saveTrace(String testTag) {
        throw new UnsupportedOperationException("Save not implemented for this monitor");
    }

    public String getChecksum() {
        return mChecksum;
    }

    static String calculateChecksum(Path traceFile) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] fileData = Files.readAllBytes(traceFile);
            byte[] hash = digest.digest(fileData);
            return BaseEncoding.base16().encode(hash).toLowerCase();
        } catch (NoSuchAlgorithmException e) {
            throw new IllegalArgumentException("Checksum algorithm SHA-256 not found", e);
        } catch (IOException e) {
            throw new IllegalArgumentException("File not found", e);
        }
    }
}

File: libraries/flicker/src/com/android/server/wm/flicker/monitor/TransitionMonitor.kt
/*
 * Copyright (C) 2020 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.server.wm.flicker.monitor

import com.android.server.wm.flicker.DEFAULT_TRACE_CONFIG
import com.android.server.wm.flicker.deleteIfExists
import com.android.server.wm.flicker.io.IResultData
import com.android.server.wm.flicker.io.ResultReader
import com.android.server.wm.flicker.io.ResultWriter
import com.android.server.wm.traces.common.ScenarioBuilder
import java.io.File

abstract class TransitionMonitor : TraceMonitor() {
    /**
     * Acquires the trace generated when executing the commands defined in the [predicate].
     *
     * @param predicate Commands to execute
     * @throws UnsupportedOperationException If tracing is already activated
     */
    fun withTracing(predicate: () -> Unit): ByteArray {
        if (this.isEnabled) {
            throw UnsupportedOperationException(
                "Trace already running. " + "This is likely due to chained 'withTracing' calls."
            )
        }
        val result: IResultData
        try {
            this.start()
            predicate()
        } finally {
            val writer = createWriter()
            this.stop(writer)
            result = writer.write()
        }
        val reader = ResultReader(result, DEFAULT_TRACE_CONFIG)
        val bytes = reader.readBytes(traceType) ?: error("Missing trace $traceType")
        result.artifact.deleteIfExists()
        return bytes
    }

    private fun createWriter(): ResultWriter {
        val className = this::class.simpleName ?: error("Missing class name for $this")
        val scenario = ScenarioBuilder().forClass(className).build()
        val tmpDir = File.createTempFile("withTracing", className).parentFile
        return ResultWriter().forScenario(scenario).withOutputDir(tmpDir)
    }
}
